### ISSUE TITLE
Don't serialise null annotations in the Plutus Core AST

### ISSUE THREAD

**Issue by kwxm** [input-output-hk/plutus#1818]
All of our AST types are parameterised over an annotation type `ann`, which is usually null.  These all get serialised and included in the CBOR, where they take up quite a lot of room: for instance, the serialised Marlowe validator is 449965 bytes long but only 354910 if you omit all the ()s.  Compression reduces these sizes to 72878 and 69715 bytes respectively, so it's not all that bad, but it's still completely redundant information.  Even after doing all the other minimisation stuff, omitting annotations saves about 10% on the final compressed size, which may be worthwhile if we really want to keep sizes down for transmission to the chain.

In `CBOR.hs` we have lots of things like 

```
instance (Serialise ann, Serialise (tyname ann), Serialise (name ann)) =>
               Serialise (Term tyname name ann)
```
My plan was to amend this to
 ```
instance (Serialise (tyname ()), serialise (name ()) => Serialise (Term tyname name ())
```
and then omit unit annotations completely from the serialisation process.  They can easily be reconstructed when we're deserialising things, by putting in `pure ()` wherever `decode` is expecting a serialised annotation.

The problem with this is that we lose the ability to serialise code that contains non-trivial annotations, but will we ever want to do that?  Also, omitting annotations maybe goes against the spirit of CBOR a bit since there's stuff in the AST that doesn't go in the CBOR, but most of the time we don't use the annotations. Note also that having non-trivial annotations will probably increase the size of serialised doce and reduce its compressibility. I originally thought that we could dispense with annotations for on-chain code entirely, but @reactormonk's costing mechanism uses annotations at run time, so we _will_ need them on chain after all, even if the transmitted code is annotation-free.

The basic problem here is that there's a default instance of `Serialise` for `()` which takes up space in the CBOR. @michaelpj suggests that we could neatly sidestep this by introducing a new type for empty annotations which has its own `Serialise` instance that does nothing.  I think this works:

```
data Empty = Empty

instance Serialise Empty where
    encode Empty = mempty
    decode = pure Empty
```

**Comment by effectfully** [comment]
Terrible nitpicking:

> `data Empty = Empty`

it should be called `Unit`, not `Empty`.

> The basic problem here is that there's a default instance of Serialise for () which takes up space in the CBOR.

Does anybody know why this is the case?

**Comment by kwxm** [comment]
> Terrible nitpicking:
> 
> > `data Empty = Empty`
> 
> it should be called `Unit`, not `Empty`.
> 

That's fine: it's just the first thing that I thought of.  

> > The basic problem here is that there's a default instance of Serialise for () which takes up space in the CBOR.
> 
> Does anybody know why this is the case?

Duncan!

I guess it's because this is supposed to be language-independent and other languages might have trouble parsing things that aren't there.  CBOR has a special tag for a null value: I think that in a language like C or JavaScript you could have an array/list of objects, some of which may be null, and you need some way to say that you've really got something there, even if it isn't a real object. 

It might even be the case that we'd be abusing CBOR by having something in the AST that isn't recorded when you serialise it.  I'd argue that annotations are just a sort of decoration that aren't really part of the AST.

**Comment by michaelpj** [comment]
It's only safe to encode unit as nothing if the presence of the unit is statically known. 

Consider `[()]` (silly example, I know). Lists in CBOR can be indefinite-length and the items are just written one after the other. So if you write unit as nothing you can't tell how many units there are in the list (maybe none).

So I think it's reasonable for the default serialization of unit to put something there - we happen to want something different.

My suggestion would be to have (~`Empty`~ `SpecialUnit`) more-or-less as an implementation detail, and then expose:

```
serializeProgram :: Program TyName Name () -> ByteString
serializeProgram p = encode $ fmap (const SpecialUnit) p

deserializeProgram :: ByteString -> Program TyName Name ()
deserializeProgram p = fmap (const ()) $ decode p
```
or something.

**Comment by effectfully** [comment]
> Consider `[()]` (silly example, I know). Lists in CBOR can be indefinite-length and the items are just written one after the other. So if you write unit as nothing you can't tell how many units there are in the list (maybe none).

That makes perfect sense.

**Comment by kwxm** [comment]
I'm not convinced that we want to hide the special unit type completely.  If you do `fmap (const ()) $ decode p` then you're possibly replacing > 100,000 secret units with real units, which will take time and space and would be wasteful on chain.  I think it'd be better to keep the special unit values in the AST until it's time to replace them with Simon's cost annotations.

**Comment by michaelpj** [comment]
I expect (but have not yet checked) that provided you do `newtype SpecialUnit = SpecialUnit ()` you should be able to implement `fmap SpecialUnit p` as `coerce`.

Also we fmap all the annotations quite a few times in various places, I'm not that worried about it unless we have evidence it's a performance problem.

**Comment by kwxm** [comment]
> Also we fmap all the annotations quite a few times in various places, I'm not that worried about it unless we have evidence it's a performance problem.

I don't care too much about what we do off chain, it's what happens on the chain that worries me: I think we should be careful about what we're doing there.  I take your point about coercion though: that may mean that this particular issue isn't something we have to worry about.

**Comment by kwxm** [comment]
Just to mention that I've been  using 

```
data U = U  -- Maybe not the best name

instance Serialise U where
    encode U = mempty
    decode = pure U
```

and then saying `U <$ prog` before serialising, and that works well and reduces the size of serialised programs substantially.  I think the thing we have to worry about is to see if we can ensure that when we do `() <$ (deserialise ...)` on chain (or _if_ we do that: we'll have to convert the AST to a form with cost annotations later anyway) we can do it without using up any time or memory; @michaelpj suggest that we can use `coerce` to arrange for `U` and `()` to have the same underlying representation, so the conversion would be zero-cost.  We should check this.

**Comment by kwxm** [comment]
Closed with #2005 (SCP-391).
